### PR TITLE
doc: SCION Packet Authenticator Option update

### PR DESCRIPTION
**Pull request by matzf**
Changes for the SCION Packet Authenticator Option specification, as
discussed in #4062.

[doc]

---
This change is [Reviewable](https://reviewable.io/reviews/scionproto/scion/4095)

### CODE REVIEW

File: doc/protocols/authenticator-option.rst [hunk 1]
```diff
@@ -5,52 +5,347 @@ SCION Packet Authenticator Option
...
+      \right)\right\rceil ms
+          = 6 ms.\\
+
+  and the absolute Unix time expressed by this Timestamp field is
+
+  .. math::
+    min_{i \in \mathrm{info\ fields}}(i.\mathrm{Timestamp}) + \mathrm{Timestamp} \cdot q
+
+  The Timestamp field can be used for replay detection by the receiver.
+  The recever MAY drop packets with timestamps outside of a locally chosen
```
Comment (marcfrei) [inline doc/protocols/authenticator-option.rst]: receiver

Comment (matzf) [reply doc/protocols/authenticator-option.rst]: Thanks, done.

**Review by mlegner — COMMENTED** [review]
Thanks for writing this up. There are a few comments below.

There are also a few minor punctuation issues, but those probably don't really matter. Also, there are some unusually long lines.

> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4095#-:-MhXqfuZ03jDlauMNVeF:b-4n980k)* status: 0 of 3 files reviewed, 12 unresolved discussions (waiting on @marcfrei and @matzf)

___
*[doc/protocols/authenticator-option.rst, line 49 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXoqbRBc5gACkQBDIP:-MhXoqbS3yqYm_vh2dre:bngay9s) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L49)):*
> ```reStructuredText
>   8-bit value 2
> OptDataLen
>   Unsigned 8-bit integer denoting the length of the full option data (1 + length of Authenticator).
> ```

This is no longer true, is it? It should be `12 + length of Authenticator`. 
Also, might be worth mentioning that this is in bytes.
___
*[doc/protocols/authenticator-option.rst, line 53 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXpErq-duF3IBtn3Ln:-MhXpErq-duF3IBtn3Lo:b7tvwdu) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L53)):*
> ```reStructuredText
> Timestamp:
>   Unsigned 24-bit integer timestamp.
>   The timestamp expressed by the value of this field is relative to the earliest `Timestamp` of
> ```

Wouldn't it be better to use the **first** info-field timestamp? That way, we would always find it in the same place instead of having to compare timestamps.
___
*[doc/protocols/authenticator-option.rst, line 67 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXpWDu2Wsi4YkTbQJH:-MhXpWDu2Wsi4YkTbQJI:bu6xv6u) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L67)):*
> ```reStructuredText
> 
>   .. math::
>     min_{i \in \mathrm{info\ fields}}(i.\mathrm{Timestamp}) + \mathrm{Timestamp} \cdot q
> ```

See previous comment.
___
*[doc/protocols/authenticator-option.rst, line 71 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXp_xK1xuimAKcfy-g:-MhXp_xK1xuimAKcfy-h:b-g7b1rp) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L71)):*
> ```reStructuredText
>   The Timestamp field can be used for replay detection by the receiver.
>   The receiver MAY drop packets with timestamps outside of a locally chosen
>   range around the current time and it MAY drop packets with duplicate
> ```

I would change this to at least "SHOULD". Also, only the tuple `source, INF timestamp, timestamp, sequence number` is actually unique.
___
*[doc/protocols/authenticator-option.rst, line 74 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXq2_W2TBEvJ3zq2M7:-MhXq2_W2TBEvJ3zq2M8:b-b497s) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L74)):*
> ```reStructuredText
>   timestamp and sequence number.
> 
> Sequence Number:
> ```

Maybe "sequence number" is not an ideal name as it suggests consecutive values. (However, I don't really have a better name...)
___
*[doc/protocols/authenticator-option.rst, line 86 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXpxhi91tafwkx7KXM:-MhXpxhi91tafwkx7KXN:b-otilcs) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L86)):*
> ```reStructuredText
>   When used with :ref:`spao-spi-drkey`, this field is used together with the
>   timestamp field to provide a unique identifier for a packet.
>   The receiver MAY drop packets with duplicate timestamp and sequence number.
> ```

See comment above.
___
*[doc/protocols/authenticator-option.rst, line 99 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXr2Tq8FRBKxYv6azo:-MhXr2Tq8FRBKxYv6azp:b-huyku8) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L99)):*
> ```reStructuredText
> Authenticator
>   This variable-length field contains the Algorithm-specific message
>   authentication code (MAC).
> ```

This could also be a combination of hash and MAC or signature.
___
*[doc/protocols/authenticator-option.rst, line 138 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXrig15DW3sQ25C1gc:-MhXrig15DW3sQ25C1gd:btd1760) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L138)):*
> ```reStructuredText
> 
> T
>   Type. Specifies the type of the key in the DRKey key hierarchy.
> ```

Could we add 1-2 more bits for changes to DRKey in the future? This would also affect the line above about the range for DRKeys (`1 \ldots 2^{19}-1` -> `1 \ldots 2^{21}-1`).
___
*[doc/protocols/authenticator-option.rst, line 161 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXsl9b2tYzmGITOvwx:-MhXsl9c37e13IgQ-AlD:b-xt6wkc) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L161)):*
> ```reStructuredText
> ==================
> 
> The authenticator for a packet is computed over the immutable fields of
> ```

Could we allow (maybe through a flag in the `Algorithm`) to also include mutable but predictable fields? This would be interesting for EPIC L3.
___
*[doc/protocols/authenticator-option.rst, line 309 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXtV6L6XvGyzNncGx9:-MhXtV6MCqWNzS9xcJvr:b-6mdubr) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L309)):*
> ```reStructuredText
> * the upper layer payload (5.)
> 
> The input to the MAC is:
> ```

I think we need to include also the address lengths from the common header. Otherwise, the length of the MAC input is not well defined and length extension attacks on CBC-MAC might be possible.

We should then also describe how length-extension attacks are prevented–i.e., by specifying the length of the input (indirectly) through the option metadata and address lengths in the first block.
___
*[doc/protocols/authenticator-option.rst, line 329 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXtyYK3hI54oScUq_7:-MhXtyYK3hI54oScUq_8:bpvjqtv) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L329)):*
> ```reStructuredText
>   - with high rates of packets (>1Gpps) we seem to need about 37 bit (~5bytes) for uniqueness
>   - timestamp should be accurate enough to allow dropping obviously old packets
>   - counter should be long enough to allow sliding window replay suppression like in IPSec
> ```

Called "sequence number" above.

**Review by matzf — COMMENTED** [review]
Thanks for the detailed feedback! Fixed most of the long lines, left a few that are just long because of references etc (following the style guide saying that readability is preferred over strict wrapping).

> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4095#-:-MhZDTs69w-z2KJVtnhP:b-48ojx4)* status: 0 of 3 files reviewed, 12 unresolved discussions (waiting on @marcfrei and @mlegner)

___
*[doc/protocols/authenticator-option.rst, line 49 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXoqbRBc5gACkQBDIP:-MhZ-f9e9O5Ucoid9g-e:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L49)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

This is no longer true, is it? It should be `12 + length of Authenticator`. 
Also, might be worth mentioning that this is in bytes.
</blockquote></details>

Done.
___
*[doc/protocols/authenticator-option.rst, line 53 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXpErq-duF3IBtn3Ln:-MhZ-ii00Kf3TBKiSnnh:ba6lk8k) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L53)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

Wouldn't it be better to use the **first** info-field timestamp? That way, we would always find it in the same place instead of having to compare timestamps.
</blockquote></details>

Ah, right, that works, good idea!
___
*[doc/protocols/authenticator-option.rst, line 67 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXpWDu2Wsi4YkTbQJH:-MhZ1vrO56SD0EbaieKW:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L67)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

See previous comment.
</blockquote></details>

Done.
___
*[doc/protocols/authenticator-option.rst, line 71 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXp_xK1xuimAKcfy-g:-MhZ227NAKu0mDCkGZmK:b-bbbkuq) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L71)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

I would change this to at least "SHOULD". Also, only the tuple `source, INF timestamp, timestamp, sequence number` is actually unique.
</blockquote></details>

You're right, SHOULD seems appropriate. I've move this snippet down to the Sequence Number field, mentioning the full tuple.
___
*[doc/protocols/authenticator-option.rst, line 74 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXq2_W2TBEvJ3zq2M7:-MhZ4fjp-cCKH1VglYiU:balr5gl) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L74)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

Maybe "sequence number" is not an ideal name as it suggests consecutive values. (However, I don't really have a better name...)
</blockquote></details>

The idea was that it would be used as an actual sequence number to do sliding-window based replay detection when this option is used with an established connection.
___
*[doc/protocols/authenticator-option.rst, line 86 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXpxhi91tafwkx7KXM:-MhZ55fg5Q3OAcGsdG71:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L86)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

See comment above.
</blockquote></details>

Done.
___
*[doc/protocols/authenticator-option.rst, line 99 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXr2Tq8FRBKxYv6azo:-MhZ5mqj3ygsfJWNPibS:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L99)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

This could also be a combination of hash and MAC or signature.
</blockquote></details>

Done.
___
*[doc/protocols/authenticator-option.rst, line 138 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXrig15DW3sQ25C1gc:-MhZCPB04Z91Kn9pWb1I:boi0ztv) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L138)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

Could we add 1-2 more bits for changes to DRKey in the future? This would also affect the line above about the range for DRKeys (`1 \ldots 2^{19}-1` -> `1 \ldots 2^{21}-1`).
</blockquote></details>

I guess it doesn't hurt, it's still less than one thousandth of the range for the SPI. Done.
___
*[doc/protocols/authenticator-option.rst, line 161 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXsl9b2tYzmGITOvwx:-MhZ61Uv7dldC4BkeWDF:b-dlwe1n) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L161)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

Could we allow (maybe through a flag in the `Algorithm`) to also include mutable but predictable fields? This would be interesting for EPIC L3.
</blockquote></details>

I'd rather not specify this without a specific use case right now, but there are enough reserved bits to add this functionality if we need it.
___
*[doc/protocols/authenticator-option.rst, line 309 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXtV6L6XvGyzNncGx9:-MhZ6YiiCuAkS3EfOFyI:b-tdgw87) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L309)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

I think we need to include also the address lengths from the common header. Otherwise, the length of the MAC input is not well defined and length extension attacks on CBC-MAC might be possible.

We should then also describe how length-extension attacks are prevented–i.e., by specifying the length of the input (indirectly) through the option metadata and address lengths in the first block.
</blockquote></details>

Yes, I struggled with this, it's just so awkward because there is no good place for this 1 byte address length/type field and we won't even make use of this, as it's skipped for DRKey Host-Host keys -- in the end I just forgot. 

Done.
___
*[doc/protocols/authenticator-option.rst, line 329 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXtyYK3hI54oScUq_7:-MhZ6vor6b6Z-e46UuYJ:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L329)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

Called "sequence number" above.
</blockquote></details>

Done.

**Review by mlegner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4095)* status: 0 of 3 files reviewed, 3 unresolved discussions (waiting on @marcfrei and @matzf)

___
*[doc/protocols/authenticator-option.rst, line 161 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXsl9b2tYzmGITOvwx:-Mhli0Ep3RMP2Hencqbx:b-mdtr8b) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L161)):*
<details><summary><i>Previously, matzf (Matthias Frei) wrote…</i></summary><blockquote>

I'd rather not specify this without a specific use case right now, but there are enough reserved bits to add this functionality if we need it.
</blockquote></details>

Maybe just add a comment for now that this is in principle possible?
___
*[doc/protocols/authenticator-option.rst, line 356 at r3](https://reviewable.io/reviews/scionproto/scion/4095#-MhliiAB7TxJ2AUv3Egw:-MhliiAB7TxJ2AUv3Egx:b-2m4q0h) ([raw file](https://github.com/scionproto/scion/blob/32c720f8485be5710899567f80c0fcf7e221a8ab/doc/protocols/authenticator-option.rst#L356)):*
> ```
> This scheme is safe from length extension attacks on the AES-CBC MAC; except
> for the addresses, all fields are of a fixed size. The length of the address
> fields is included in the first block of the AES-CBC MAC.
> ```
An crucially, if the addresses and their length is not included, this fact can be seen from the metadata, so also in that case no extension attacks are possible.

**Review by mlegner — COMMENTED** [review]
Thanks for the changes. Two minor points left. :)

> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4095#-:-MhljOLaB_UBaEs3rT4o:bsiklmx)* status: 0 of 3 files reviewed, 3 unresolved discussions (waiting on @marcfrei and @matzf)

**Comment by marcfrei** [comment]
___
*[doc/protocols/authenticator-option.rst, line 62 at r3](https://reviewable.io/reviews/scionproto/scion/4095#-Miu54LS4HXS1Zn3j2QY:-Miu54LTBT5kwU3yHksH:bea3puu) ([raw file](https://github.com/scionproto/scion/blob/32c720f8485be5710899567f80c0fcf7e221a8ab/doc/protocols/authenticator-option.rst#L62)):*
> ```reStructuredText
>       q := \left\lceil\left(
>         \frac{24 \times 60 \times 60 \times 10^3}
>              {2^{32}}
> ```

Not sure, but shouldn't this be {2^{24}} in the denominator?

**Comment by marcfrei** [comment]
___
*[doc/protocols/authenticator-option.rst, line 115 at r1](https://reviewable.io/reviews/scionproto/scion/4095#-MiuUaJ55Y_oZY-TrbuY:-MiuUaJ62rB2PdOGGtH5:brrz8se) ([raw file](https://github.com/scionproto/scion/blob/bca2b6755a185a66c163a744896fab732ea580b9/doc/protocols/authenticator-option.rst#L115)):*
> ```reStructuredText
> and MUST NOT be sent on the wire.
> 
> SPI values in the range :math:`1 \ldots 2^{19}-1` identify a DRKey.
> ```

Out of curiosity, why 2^{19}?

**Comment by marcfrei** [comment]
___
*[doc/protocols/authenticator-option.rst, line 340 at r1](https://reviewable.io/reviews/scionproto/scion/4095#-Miv994iCi0345x31AF0:-Miv994iCi0345x31AF1:b-4tw7e3) ([raw file](https://github.com/scionproto/scion/blob/bca2b6755a185a66c163a744896fab732ea580b9/doc/protocols/authenticator-option.rst#L340)):*
> ```reStructuredText
> 
> - this does not appear to work with less than 3 rows. We use the available
>   room to make the timestamp and counter 3 bytes each and leave on reserved
> ```

`one reserved byte`

**Comment by marcfrei** [comment]
___
*[doc/protocols/authenticator-option.rst, line 351 at r1](https://reviewable.io/reviews/scionproto/scion/4095#-Miv9cm51gsGNjv1g760:-Miv9cm65O2f2PZxhWzr:b-qp5mg2) ([raw file](https://github.com/scionproto/scion/blob/bca2b6755a185a66c163a744896fab732ea580b9/doc/protocols/authenticator-option.rst#L351)):*
> ```reStructuredText
>     this enough?
>     Then 3 byte SPI + 2 byte counter + 3 byte timestamp gives a horribly
>     misaligned mess but it would be only two rows.
> ```

I would probably remove the last point above about a possible alternative approach.

**Comment by marcfrei** [comment]
___
*[doc/protocols/authenticator-option.rst, line 335 at r1](https://reviewable.io/reviews/scionproto/scion/4095#-MivCTjn9YK5fvL9KPEY:-MivCTjo5sbye-CBMNX2:b-3gcgvk) ([raw file](https://github.com/scionproto/scion/blob/bca2b6755a185a66c163a744896fab732ea580b9/doc/protocols/authenticator-option.rst#L335)):*
> ```reStructuredText
> - reasonable field alignment with little padding with 4n + 2 option alignment (to avoid padding before first option)
> 
> - 2 AES blocks or fewer for lightning filter usecase (SHA1-AES-CBC with DRKey)
> ```

"use case" instead of "usecase", like in the next point.

**Comment by marcfrei** [comment]
___
*[doc/protocols/authenticator-option.rst, line 79 at r3](https://reviewable.io/reviews/scionproto/scion/4095#-MivDL_VCwvs-ySJAFl6:-MivDL_W3eAUHgrrwUuM:b-cr8ytd) ([raw file](https://github.com/scionproto/scion/blob/32c720f8485be5710899567f80c0fcf7e221a8ab/doc/protocols/authenticator-option.rst#L79)):*
> ```reStructuredText
>   This field can be used for replay detection by the receiver.
> 
>   The tuple
> ```

Left-ver from editing?

**Review by matzf — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4095)* status: 0 of 3 files reviewed, 6 unresolved discussions (waiting on @marcfrei and @mlegner)

___
*[doc/protocols/authenticator-option.rst, line 115 at r1](https://reviewable.io/reviews/scionproto/scion/4095#-MiuUaJ55Y_oZY-TrbuY:-Mj4_R0ZDKGS38nqxzln:bla7ooo) ([raw file](https://github.com/scionproto/scion/blob/bca2b6755a185a66c163a744896fab732ea580b9/doc/protocols/authenticator-option.rst#L115)):*
<details><summary><i>Previously, marcfrei (Marc Frei) wrote…</i></summary><blockquote>

Out of curiosity, why 2^{19}?
</blockquote></details>

It's been changed to 2^21 now (reserved two more bits). It's just the number of bits used used by the encoding defined below to identify a specific DRKey.
___
*[doc/protocols/authenticator-option.rst, line 335 at r1](https://reviewable.io/reviews/scionproto/scion/4095#-MivCTjn9YK5fvL9KPEY:-Mj4gaNH7S7zpueE5jBm:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/bca2b6755a185a66c163a744896fab732ea580b9/doc/protocols/authenticator-option.rst#L335)):*
<details><summary><i>Previously, marcfrei (Marc Frei) wrote…</i></summary><blockquote>

"use case" instead of "usecase", like in the next point.
</blockquote></details>

Done.
___
*[doc/protocols/authenticator-option.rst, line 340 at r1](https://reviewable.io/reviews/scionproto/scion/4095#-Miv994iCi0345x31AF0:-Mj4h-RNAc_WGkokJX6r:b-tu528l) ([raw file](https://github.com/scionproto/scion/blob/bca2b6755a185a66c163a744896fab732ea580b9/doc/protocols/authenticator-option.rst#L340)):*
<details><summary><i>Previously, marcfrei (Marc Frei) wrote…</i></summary><blockquote>

`one reserved byte`
</blockquote></details>

I don't see it. Is there a typo or is it just wrong?
___
*[doc/protocols/authenticator-option.rst, line 351 at r1](https://reviewable.io/reviews/scionproto/scion/4095#-Miv9cm51gsGNjv1g760:-Mj4hARfDEK1HR04Vj3S:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/bca2b6755a185a66c163a744896fab732ea580b9/doc/protocols/authenticator-option.rst#L351)):*
<details><summary><i>Previously, marcfrei (Marc Frei) wrote…</i></summary><blockquote>

I would probably remove the last point above about a possible alternative approach.
</blockquote></details>

Done.
___
*[doc/protocols/authenticator-option.rst, line 161 at r2](https://reviewable.io/reviews/scionproto/scion/4095#-MhXsl9b2tYzmGITOvwx:-Mj4cAPX-QMtVkhpi4sJ:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/fffdc75030ee3b910d804ecc96a723df6d0916fa/doc/protocols/authenticator-option.rst#L161)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

Maybe just add a comment for now that this is in principle possible?
</blockquote></details>

Done.
___
*[doc/protocols/authenticator-option.rst, line 62 at r3](https://reviewable.io/reviews/scionproto/scion/4095#-Miu54LS4HXS1Zn3j2QY:-Mj4Z3vQ14TtJCiUaoBt:b-kwf5zq) ([raw file](https://github.com/scionproto/scion/blob/32c720f8485be5710899567f80c0fcf7e221a8ab/doc/protocols/authenticator-option.rst#L62)):*
<details><summary><i>Previously, marcfrei (Marc Frei) wrote…</i></summary><blockquote>

Not sure, but shouldn't this be {2^{24}} in the denominator?
</blockquote></details>

:100: Done.
___
*[doc/protocols/authenticator-option.rst, line 79 at r3](https://reviewable.io/reviews/scionproto/scion/4095#-MivDL_VCwvs-ySJAFl6:-Mj4_NUf4cjWKXLM3zYj:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/32c720f8485be5710899567f80c0fcf7e221a8ab/doc/protocols/authenticator-option.rst#L79)):*
<details><summary><i>Previously, marcfrei (Marc Frei) wrote…</i></summary><blockquote>

Left-ver from editing?
</blockquote></details>

Done.
___
*[doc/protocols/authenticator-option.rst, line 356 at r3](https://reviewable.io/reviews/scionproto/scion/4095#-MhliiAB7TxJ2AUv3Egw:-MjDa2iWCpi5tAn8C9UB:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/32c720f8485be5710899567f80c0fcf7e221a8ab/doc/protocols/authenticator-option.rst#L356)):*
<details><summary><i>Previously, mlegner (Markus Legner) wrote…</i></summary><blockquote>

> ```
> This scheme is safe from length extension attacks on the AES-CBC MAC; except
> for the addresses, all fields are of a fixed size. The length of the address
> fields is included in the first block of the AES-CBC MAC.
> ```
An crucially, if the addresses and their length is not included, this fact can be seen from the metadata, so also in that case no extension attacks are possible.
</blockquote></details>

Done.

**Review by matzf — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4095)* status: 0 of 3 files reviewed, 6 unresolved discussions (waiting on @marcfrei and @mlegner)

___
*[doc/protocols/authenticator-option.rst, line 340 at r1](https://reviewable.io/reviews/scionproto/scion/4095#-Miv994iCi0345x31AF0:-MjDbcKdEUUYQ2BYnKl7:bdaqxj9) ([raw file](https://github.com/scionproto/scion/blob/bca2b6755a185a66c163a744896fab732ea580b9/doc/protocols/authenticator-option.rst#L340)):*
<details><summary><i>Previously, matzf (Matthias Frei) wrote…</i></summary><blockquote>

I don't see it. Is there a typo or is it just wrong?
</blockquote></details>

Aah gotcha, it was a typo (i had `on reserved byte`) and I seem to have fixed it and forgotten to mark this as done.
Done.

**Comment by marcfrei** [comment]
:lgtm:

**Comment by fstreun** [comment]
___
*[doc/protocols/authenticator-option.rst, line 356 at r5](https://reviewable.io/reviews/scionproto/scion/4095#-Mkgo8V25Gl-5PY2pQpf:-Mkgo8V25Gl-5PY2pQpg:b-2wgmsc) ([raw file](https://github.com/scionproto/scion/blob/aa550f92b1a7d48a52e7ffda86ae8e17b8c22b19/doc/protocols/authenticator-option.rst#L356)):*
> ```
> Observe that when used with suitable an :ref:`SPI referring to a DRKey
> <spao-spi-drkey>`, the address header may be left empty, resulting in an ideal
> ```

There seems a 'an' to be at the wrong place.
Maybe:
Observe that when used with a suitable [...]

**Comment by fstreun** [comment]
___
*[doc/protocols/authenticator-option.rst, line 337 at r5](https://reviewable.io/reviews/scionproto/scion/4095#-MkgulRD475NmNXpBKxH:-MkgulRE994ziFy5xHX6:b-pczwop) ([raw file](https://github.com/scionproto/scion/blob/aa550f92b1a7d48a52e7ffda86ae8e17b8c22b19/doc/protocols/authenticator-option.rst#L337)):*
> ```
> Adress
> ```
Address

**Review by oncilla — APPROVED** [review]
:lgtm:

> Reviewed 2 of 3 files at r1, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4095#-:-MoF4L9224xm9CYIKf3B:bnfp4nl)* status: all files reviewed, 4 unresolved discussions (waiting on @matzf and @mlegner)

**Review by matzf — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4095)* status: 1 of 3 files reviewed, 3 unresolved discussions (waiting on @fstreun, @mlegner, and @oncilla)

___
*[doc/protocols/authenticator-option.rst, line 337 at r5](https://reviewable.io/reviews/scionproto/scion/4095#-MkgulRD475NmNXpBKxH:-MoI9jwf2shUP1aOJLF8:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/aa550f92b1a7d48a52e7ffda86ae8e17b8c22b19/doc/protocols/authenticator-option.rst#L337)):*
<details><summary><i>Previously, fstreun wrote…</i></summary><blockquote>

> ```
> Adress
> ```
Address
</blockquote></details>

Done.
___
*[doc/protocols/authenticator-option.rst, line 356 at r5](https://reviewable.io/reviews/scionproto/scion/4095#-Mkgo8V25Gl-5PY2pQpf:-MoIA4vK6GnsZVn8ea7Q:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/aa550f92b1a7d48a52e7ffda86ae8e17b8c22b19/doc/protocols/authenticator-option.rst#L356)):*
<details><summary><i>Previously, fstreun wrote…</i></summary><blockquote>

> ```
> Observe that when used with suitable an :ref:`SPI referring to a DRKey
> <spao-spi-drkey>`, the address header may be left empty, resulting in an ideal
> ```

There seems a 'an' to be at the wrong place.
Maybe:
Observe that when used with a suitable [...]
</blockquote></details>

Done.

**Review by oncilla — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4095)* status: 1 of 3 files reviewed, 3 unresolved discussions (waiting on @fstreun, @mlegner, and @oncilla)

*[a discussion](https://reviewable.io/reviews/scionproto/scion/4095#-MoIXdJS4oT1np6JztYH:-MoIXdJT3bUh-JW21jq-:bus1eau) (no related file):*
@matzf I was too trigger happy, in addition to https://github.com/scionproto/scion/pull/4127 there will be another commit that takes your latest commit into conisderation.

---

**Review by matzf — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4095)* status: 1 of 3 files reviewed, 3 unresolved discussions (waiting on @fstreun, @mlegner, and @oncilla)

*[a discussion](https://reviewable.io/reviews/scionproto/scion/4095#-MoIXdJS4oT1np6JztYH:-MoIYANcBafv386JofBu:b-i5qi2o) (no related file):*
<details><summary><i>Previously, oncilla (Dominik Roos) wrote…</i></summary><blockquote>

@matzf I was too trigger happy, in addition to https://github.com/scionproto/scion/pull/4127 there will be another commit that takes your latest commit into conisderation.
</blockquote></details>

Ah, ok no problem for me. Sorry, for causing a mess, I had forgotten to fix these small typos until your review.

---

**Review by mlegner — COMMENTED** [review]
> Reviewed 1 of 2 files at r6.
> *[Reviewable](https://reviewable.io/reviews/scionproto/scion/4095)* status: 2 of 3 files reviewed, 1 unresolved discussion (waiting on @fstreun and @oncilla)

**Pull request closed without merging**